docs/Changelog.md

```diff
@@ -18040,19 +18040,37 @@ This version of the operator has been available since version 13 of the default
 ### <a name="Slice-13"></a>**Slice-13**</a>
 
   Produces a slice of the input tensor along multiple axes. Similar to numpy:
-  https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html
-  Slices uses `starts`, `ends`, `axes` and `steps` inputs to specify the start and end
-  dimension and step for each axis in the list of axes, it uses this information to
-  slice the input `data` tensor. If a negative value is passed for any of the
-  start or end indices, it represents number of elements before the end of that
-  dimension. If the value passed to start or end is larger than the `n` (the
-  number of elements in this dimension), it represents `n`. For slicing to the
-  end of a dimension with unknown size, it is recommended to pass in `INT_MAX`
-  when sclicing forward and 'INT_MIN' when slicing backward.
-  If a negative value is passed for step, it represents slicing backward.
-  However step value cannot be 0.
-  If `axes` are omitted, they are set to `[0, ..., ndim-1]`.
+  https://numpy.org/doc/stable/user/basics.indexing.html?highlight=slice#slicing-and-striding
+
+  Slice uses the `starts`, `ends`, `axes` and `steps` inputs to select a sub-tensor
+  of its input `data` tensor.
+
+  An effective `start[i]`, `end[i]`, and `step[i]` must be computed for each `i`
+  in `[0, ... r-1]` where `r = rank(input)` as follows:
+
+  If `axes` are omitted, they are set to `[0, ..., r-1]`.
   If `steps` are omitted, they are set to `[1, ..., 1]` of length `len(starts)`
+
+  The effective values are initialized as `start[i] = 0`, `end[i] = dims[i]` where
+  `dims` are the dimensions of `input` and `step[i] = `1.
+
+  All negative elements of `axes` are made non-negatve by adding `r` to them, where
+  `r =rank(input)`.
+
+  All negative values in `starts[i]` and `ends[i]` have `dims[axes[i]]` added to them,
+  where `dims` are the dimensions of `input`. Then `start[axes[i]]` is the adjusted
+  `starts[i]` clamped into range of valid indices, i.e. `[0, dims[axes[i]]-1]`.
+
+  The clamping for the adjusted `ends[i]` depends on the sign of `steps[i]` and must
+  accommodate copying 0 through `dims[axes[i]]` elements, so for positive stepping
+  `end[axes[i]]` is clamped to `[0, dims[axes[i]]]`, while for negative stepping it
+  is clamped to `[-1, ends[i]-1]`.
```

Comment:
Should this read `dims[axes[i]]-1` or am I mis-understanding something ?

Reply:
Starts are inclusive, ends are exclusive, so the slice covers indices `{i | start <= i < end}`.
If it still seems wrong or you see a way to make it clearer, let me know.

Reply:
Sorry for not being clear enough. I think the logic is right. Looking at it, it seems inconsistent with the way it is `worded`. 

`...so for positive stepping `ends[axes[i]]` is clamped to `[0, dims[axes[i]]]`

Looking at the above line, I would have thought, the last set of words should be `.... is clamped to [-1, dims[axes[i]] - 1]` instead of the `ends` you are referencing. Does this make sense or am I on the wrong track ? :)

Reply:
Let's say we have a tensor X[5]. Then a `slice(X, start=0, end=0)` is empty, `slice(X, start=0, end=5)` is all of X, and `slice(X, start=0, end=101086)` is the same as `slice(X, start=0, start=5)`.
In ranges, `[` and `]` are inclusive, `(` and `)` are exclusive.

Reply:
I agree with everything above. 

I was just pointing out the clamping wording while negative stepping:
`[-1, ends[i]-1]`   ->   `[-1, dims[axes[i]] - 1]` ?

Reply:
@hariharans29 : I agree with your comment about the wording, which seems like a typo. More importantly, I was wondering about the issue you created https://github.com/onnx/onnx/issues/4115 ... do you still have a concern about this? It seemed like an important point, but I haven't checked it out yet.

Reply:
Thanks Rama, yes, it seems like a typo.

Re: #4115 - To have the link to Numpy's slicing in the op's description and prescribe a different clamping for `start` is a bit misleading. ORT implements the Numpy clamping (which does the clamping based on the slicing direction) for `start` since the time `steps` input was introduced in Slice-10. This is because the spec initially didn't have the clamping logic called out explicitly and I used Numpy's clamping as the de-facto (Numpy's Slice was also linked to in the op's spec).  FWIW- The "real-world" impact of this should be minimal. It is just that ORT will need to make changes if it needs to "honor" this newly called out clamping logic. As much as this discussion is more pedantic than a burning issue, it would be nice to clarify this.

Reply:
I hadn't seen #4115 and thought you were referring to the positive stepping case. Your change looks correct and agrees with what I implemented in our glow fork (although I used the equivalent adjusted start instead of `dim[axes[i]]-1`, so that's two implementations.

Reply:
The original textual description was unclear. However, how about the original shape-inference code?
(a) Is the original shape-inference code consistent with the ORT/glow implementation or not?
(b) Does the change in shape-inference code in this PR change the behavior or does it not? This line of change seems relevant to the kind of counterexample that Hari discusses in the above issue, but I feel there may be edge-cases where the shape-inference code may not be consistent, and it would be helpful to check this.

Reply:
This was the original shape inference code:
 
 if (step < 0)
    start = clamp(start, 0, input_rank - 1);
  else
    start = clamp(start, 0, input_rank);

This is how ORT implements it: https://github.com/microsoft/onnxruntime/blob/26fceca90f35fe94f39864ebaec0f2d51e367f10/onnxruntime/core/providers/cpu/tensor/slice_helper.h#L132

So, the old shape inference code matches with ORT's implementation.

Reply:
So, I think we should revert to the old shape-inference code. Is there any case where the old shape-inference code is incorrect?

Reply:
And change the textual description accordingly. (Since we are not bumping the opset version, this PR should be only a clarification, not a change in behavior.)

Reply:
Does that sound reasonable? Thanks!

Reply:
I'm not seeing how the step > 0 case above is correct; start is inclusive, which is [0, input_rank), not [0, input_rank+1].

Reply:
NumPy does behave like the original.

Reply:
@gramalingam : I think reverting to the old shape inference will be a good idea. It atleast maintains status quo and as you rightly say changing any behavior without an opset bump up is not a good idea.

@diyessi : Try this example on Numpy (this is for step > 0),
x[5:6:1]
array([], dtype=int32)

If `start` had been clamped to an inclusive value as you are suggesting, with the `ends` clamping the way we all agree it to be, the result would actually be non-empty